src/query.rs

```diff
@@ -65,6 +65,8 @@ impl From<ReleaseType> for ActualSemverUpdate {
 #[non_exhaustive]
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct SemverQuery {
+    // this should be populated with the file name and not serialized
+    #[serde(skip)]
```

Comment:
I've never tried to combine `serde(skip)` of a non-`Option` field together with a `#[derive(Deserialize)]`.

Do you know what happens if someone tries to deserialize a `SemverQuery` type using that derived impl? I genuinely have no idea what the result would be, and I'm worried we might be leaving a footgun here.

Reply:
According to serde's [docs](https://serde.rs/field-attrs.html#skip), `#[serde(skip)]` should unconditionally fill it in with `Default::default()` (i.e. `String::new()`). It hadn't occurred to me before that this would constitute a change in behavior on a publicly visible field. 

Perhaps it would be better to use [`#[serde(default)]`](https://serde.rs/field-attrs.html#default) instead of skip. This would still deserialize the field if it exists just like before, the only observable change would be that if the `id` field is missing, deserialization attempts will now succeed whereas before they would return an error. It's conceivable that could break someone's workflow if they have their own tests relying on deserialization without id to fail but that seems unlikely to me.

Another option is to not touch the serde attributes on the struct at all. This would require putting the id field back in the lints/*.ron files, but we can use a throwaway placeholder value to indicate that it will be overwritten (with the filename derived id).

Reply:
Taking a couple of steps back here: the underlying problem I think we're solving is the possibility that someone writing / editing a lint file might set the `id` incorrectly. Right?

The recommended workflow for making new lints is to use `./scripts/make_new_lint.sh`, which automatically fills in the `id` field. In that case, mistakes are unlikely: they'd require an explicit change to the `id` field.

If we apply either `serde` attribute, or override the field's value on deserialization, that seems to introduce quite a bit of non-obvious behavior and risks being a footgun.

An alternative option would be to add a test that enforces that all lint file names match the `id` string in the file. Normally, this would be annoying since new contributors might not know about this requirement or about the test. But we expect new lints to be created using the above workflow that automatically sets this field correctly, so this test should in practice never fail for new contributors _unless_ they mistakenly edited the `id` field by hand.

I'm inclined to think that adding such a test is the lesser inconvenience here, rather than a fancier solution with `serde` attributes or other magic that might backfire.

What do you think?